oqsprov/oqs_encode_key2any.c

```diff
@@ -566,89 +640,262 @@ static int oqsx_pki_priv_to_der(const void *vxkey, unsigned char **pder)
+                }
+
+                if (sk_ASN1_TYPE_num(sk) != -1)
+                    sk_ASN1_TYPE_pop_free(sk, &ASN1_TYPE_free);
+                else
+                    ASN1_TYPE_free(aType[i]);
+
+                OPENSSL_free(aType);
+                OPENSSL_free(aString);
+                OPENSSL_free(temp);
+                OPENSSL_free(templen);
+                PKCS8_PRIV_KEY_INFO_free(p8inf_internal);
+                OPENSSL_free(name);
+                return -1;
+            }
+
+            if (get_oqsname_fromtls(name) == 0) {
+
+                nid = oqsxkey->oqsx_provider_ctx.oqsx_evp_ctx->evp_info
+                          ->keytype;
+                if (nid == EVP_PKEY_RSA) { // get the RSA real key size
+                    unsigned char *enc_len
+                        = OPENSSL_strndup(oqsxkey->comp_privkey[i], 4);
+                    OPENSSL_cleanse(enc_len, 2);
+                    DECODE_UINT32(buflen, enc_len);
+                    buflen += 4;
+                    OPENSSL_free(enc_len);
+                    if (buflen > oqsxkey->privkeylen_cmp[i]) {
+                        for (int j = 0; j <= i; j++) {
+                            OPENSSL_cleanse(aString[j]->data,
+                                            aString[j]->length);
+                            ASN1_OCTET_STRING_free(aString[j]);
+                            OPENSSL_cleanse(aType[j]->value.sequence->data,
+                                            aType[j]->value.sequence->length);
+                            if (j < i)
+                                OPENSSL_clear_free(temp[j], templen[j]);
+                        }
+
+                        if (sk_ASN1_TYPE_num(sk) != -1)
+                            sk_ASN1_TYPE_pop_free(sk, &ASN1_TYPE_free);
+                        else
+                            ASN1_TYPE_free(aType[i]);
+
+                        OPENSSL_free(aType);
+                        OPENSSL_free(aString);
+                        OPENSSL_free(temp);
+                        OPENSSL_free(templen);
+                        PKCS8_PRIV_KEY_INFO_free(p8inf_internal);
+                        OPENSSL_free(name);
+                        return -1;
+                    }
+                } else
+                    buflen = oqsxkey->privkeylen_cmp[i];
+            } else {
+                nid = OBJ_sn2nid(name);
+                buflen = oqsxkey->privkeylen_cmp[i] + oqsxkey->pubkeylen_cmp[i];
+            }
 
-    OPENSSL_secure_clear_free(buf, buflen);
+            buf = OPENSSL_secure_malloc(buflen);
+            if (get_oqsname_fromtls(name)
+                != 0) { // include pubkey in privkey for PQC
+                memcpy(buf, oqsxkey->comp_privkey[i],
+                       oqsxkey->privkeylen_cmp[i]);
+                memcpy(buf + oqsxkey->privkeylen_cmp[i],
+                       oqsxkey->comp_pubkey[i], oqsxkey->pubkeylen_cmp[i]);
+            } else {
+                memcpy(buf, oqsxkey->comp_privkey[i], buflen);
```

Comment:
I accept it because at first glance it seems fine, but then I went to test it and remembered why I did that.
Some cases of the RSA keys the size of the key will not match the size on oqsxkey->privkeylen_cmp, this value represent the maximum size for the RSA, so I cant copy oqsxkey->privkeylen_cmp because it will cause overflow in some cases.